helix-core/src/case_conversion.rs

```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
+    separator: Option<char>,
+) {
+    let mut capitalize_next = capitalize_first;
+    let mut prev: Option<char> = None;
+
+    for c in text.skip_while(|ch| ch.is_whitespace()) {
+        if c.is_alphanumeric() {
+            if prev.is_some_and(|p| p.is_lowercase()) && c.is_uppercase() {
+                capitalize_next = true;
+            }
+            if capitalize_next {
+                buf.push(c.to_ascii_uppercase());
+                capitalize_next = false;
+            } else {
+                buf.extend(c.to_lowercase());
+            }
         } else {
-            buf.push(c)
+            capitalize_next = true;
+            if let Some(separator) = separator {
+                if prev.is_some_and(|p| p != separator) {
+                    buf.push(separator);
+                }
+            }
         }
+        prev = Some(c);
     }
 }
 
-pub fn to_upper_case_with(text: impl Iterator<Item = char>, buf: &mut Tendril) {
-    for c in text {
-        for c in c.to_uppercase() {
-            buf.push(c)
+pub fn separator_case_conversion(
+    text: impl Iterator<Item = char>,
+    buf: &mut Tendril,
+    separator: char,
+) {
+    let mut prev: Option<char> = None;
+
+    for c in text.skip_while(|ch| ch.is_whitespace()) {
+        if c.is_alphanumeric() {
+            if prev.is_some_and(|p| p.is_lowercase()) && c.is_uppercase()
+                || !prev.is_some_and(|p| p.is_alphanumeric()) && !buf.is_empty()
+            {
+                buf.push(separator);
+            }
+
+            buf.push(c.to_ascii_lowercase());
         }
+        prev = Some(c);
     }
 }
 
-pub fn to_lower_case_with(text: impl Iterator<Item = char>, buf: &mut Tendril) {
-    for c in text {
-        for c in c.to_lowercase() {
-            buf.push(c)
+pub fn into_alternate_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, |c| {
+        if c.is_uppercase() {
+            c.to_ascii_lowercase()
+        } else if c.is_lowercase() {
+            c.to_ascii_uppercase()
+        } else {
+            *c
         }
-    }
+    });
 }
 
-pub fn to_camel_case(text: impl Iterator<Item = char>) -> Tendril {
+pub fn into_upper_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, char::to_ascii_uppercase);
+}
+
+pub fn into_lower_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    simple_case_conversion(text, buf, char::to_ascii_lowercase);
+}
+
+pub fn into_kebab_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    separator_case_conversion(text, buf, '-');
+}
+
+pub fn into_snake_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    separator_case_conversion(text, buf, '_');
+}
+
+pub fn into_title_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    complex_case_conversion(text, buf, true, Some(' '));
+}
+
+pub fn into_camel_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
```

Comment:
Since we have Title Case we should probably include Sentence case as well

Reply:
I don't think it's very useful. Maybe I'm wrong, but also `s` is reserved for snake_case and what key would be reserved for sentence case? S wouldn't make sense since usually a lowercase letter and capital letter will do opposite actions to each other like `p` and `P`. So we would have `s` for snake_case and `S` for sentence case, when they're not really related 

I just think consistency is important. but if you have some good ideas let me know!